src/System.Runtime.Numerics/src/System/Numerics/BigCalc.SquMul.cs

```diff
@@ -0,0 +1,356 @@
+﻿using System.Diagnostics;
+using System.Security;
+
+namespace System.Numerics
+{
+    internal static partial class BigCalc
+    {
+#if DEBUG
+
+        private const int SquareThreshold = 128 / 32;
+        private const int MultiplyThreshold = 128 / 32;
+
+#else
+
+        private const int SquareThreshold = 1024 / 32;
+        private const int MultiplyThreshold = 1024 / 32;
+
+#endif
+
+        public static uint[] Square(uint[] value)
+        {
+            Debug.Assert(value != null);
+
+            return Square(value, value.Length);
+        }
+
+        [SecuritySafeCritical]
+        private unsafe static uint[] Square(uint[] value, int valueLength)
+        {
+            Debug.Assert(value != null);
+            Debug.Assert(value.Length >= valueLength);
+
+            var bits = new uint[valueLength * 2];
+
+            fixed (uint* v = value, b = bits)
+            {
+                Square(v, valueLength,
+                       b, bits.Length);
+            }
+
+            return bits;
+        }
+
+        [SecuritySafeCritical]
+        private unsafe static void Square(uint* value, int valueLength,
+                                          uint* bits, int bitsLength)
+        {
+            Debug.Assert(valueLength >= 0);
+            Debug.Assert(bitsLength == valueLength * 2);
+
+            if (valueLength < SquareThreshold)
+            {
+                // squares the bits
+                for (var i = 0; i < valueLength; i++)
+                {
+                    var carry = 0UL;
+                    for (var j = 0; j < i; j++)
+                    {
+                        var digit1 = bits[i + j] + carry;
+                        var digit2 = (ulong)value[j] * value[i];
```

Comment:
It might be better to not use "var" in this code. It would be easier for me to see what is going on with the conversions between uints and ulongs.